addons/website_slides_survey/models/survey_user.py

```diff
@@ -11,3 +11,40 @@ class SurveyUserInput(models.Model):
         help="The related course slide when there is no membership information")
     slide_partner_id = fields.Many2one('slide.slide.partner', 'Subscriber information',
         help="Slide membership information for the logged in user")
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        records = super(SurveyUserInput, self).create(vals_list)
+        records._check_for_failed_attempt()
+        return records
+
+    @api.multi
+    def write(self, vals):
+        res = super(SurveyUserInput, self).write(vals)
+        self._check_for_failed_attempt()
+        return res
+
+    def _check_for_failed_attempt(self):
+        """ If the user fails his last attempt at a course certification,
+        we remove him from the members of the course (and he has to enroll again).
+        He receives an email in the process notifying him of his failure and suggesting
+        he enrolls to the course again.
+
+        The purpose is to have a 'certification flow' where the user can re-purchase the
+        certification when they have failed it."""
+        for record in self:
```

Comment:
I think we can make this a bit more in batch, like

1/ filter user_inputs (maybe with a search to lessen number of records to prefetch)
2/ compute failed user inputs
3/ remove partnership in batch and send mails (loop required as it is a single id method, without force_send)

Reply:
I made an attempt at this in a separated commit.
Code looks more complicated even though it's now batching the deletes.

Maybe you can see a more effective/easier to read way of doing it?